<H1 style="font-size:50px">3: Tipi di Dati e Data Structures</H1>

<div style = "width:95%; padding:5px 10px 5px;  margin:50px 0px 10px 10px; font-size:18px">
<UL>
    <LI> Quali sono i principali tipi numerici di dati
    <LI> Quali sono i principali tipi di dati sequenziali 
    <UL> 
        <LI> Le stringhe di testo
        <LI> Le liste
        <LI> Le ntuple
    </UL>
    <LI> Come utilizzare gli indici.
    <LI> Come estrerre sezioni di sequenze
    <LI> Cosa sono i dizionari di Python
    <LI> Come convertire un tipo in un altro
    <LI> Come le variabili vengono immagazzinate in memoria. Tipi mutabili e immutabili.
    <LI> Uguaglianza e identità fra variabili
</UL>
</div>

<hr>

Python conosce vari tipi di dati. Il tipo di una variabile viene ricavato dal valore che le viene assegnato. Nei linguaggi in cui un programma deve essere compilato prima di essere eseguito il tipo di ciascuna variabile deve essere dichiarato prima di assegnare alla variabile un valore.
Python è un linguaggio interpretato i comandi vengono letti ed eseguiti sequenzialmente.

3.1 Dati numerici
---------

I tipi numerici built-in sono i numeri interi (integers, int), che possono essere anche rappresentati in base esadecimale, ottale o binaria, i numeri reali (floating point, float) e quelli complessi.

### Numeri interi

Una successione di cifre che non contenga il punto di separazione viene interpretato come un numero intero. 

In [2]:
35**42

70934557307860443711736098025989133248003781773149967193603515625

### Numeri esadecimali

Sono numeri interi in base 16. Le cifre sono 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, e, f.

In [5]:
a = 0xf
int(a)

15

In [3]:
b = 1357
hex(b)

'0x54d'

### Numeri reali

Una successione di cifre che contenga il punto di separazione viene interpretato come un numero reale.
__I numeri reali sono rappresentati in modo approssimato (10-14 cifre di precisione).__

In [ ]:
0.3

In [1]:
0.1*3

0.30000000000000004

Numeri reali in notazione scientifica: $1.3\, 10^{-4}$ si scrive 1.3e-4.

In [15]:
a = 9.7e-4
b = 3.0e+17
a*b

'2.910000e+14'

Python, di default, scrive i numeri reali in formato non scientifico. Se vogliamo la risposta in notazione scientifica è necessario chiederlo esplicitamente (Maggiori informazioni più avanti.).

In [6]:
f"{a*b:e}"   # f"{n°:e}"

'2.790000e+14'

In [1]:
f"{91000000:e}"

'9.100000e+07'

### Numeri complessi

In Python (come in Fortran e Matlab) i numeri complessi sono built-in. Qualche esempio:

In [2]:
x = 1.2 + 3.7j     # Nessuno spazio fra il valore della parte immaginaria e il simbolo `j`
x

(1.2+3.7j)

In [3]:
x.imag  # Parte immaginaria

3.7

In [4]:
x.real  # Parte reale

1.2

In [5]:
x * x.conjugate()

(15.13+0j)

La sintassi per le operazioni fra numeri complessi è identica a quella per le operazioni fra numeri reali:

In [6]:
x * x

(-12.250000000000002+8.88j)

In [7]:
3 * x

(3.5999999999999996+11.100000000000001j)

Se si utilizza una variabile per la parte immaginaria, la notazione per costruire un numero complesso è: 

In [8]:
a = 1.
b = -0.5
c = a + b*1j
print(c)

(1-0.5j)


Per eseguire operazioni più complicate (come fare radici quadrate di numeri complessi etc) è nessario usare il modulo `cmath` (Complex MATHematics):

In [9]:
import cmath
cmath.sqrt(-1)

1j

### Funzioni che si  applicano a tutti i tipi di numeri

La funzione `abs()`  restituisce il valore assoluto o modulo di un numero:

In [10]:
a = -45.463
abs(a)

45.463

`abs()` funziona anche per numeri complessi.

In [11]:
b = abs(1 + 2j)
b

2.23606797749979

In [12]:
b*b

5.000000000000001

<div style = "width:95%; padding:5px 10px 5px; border-style:solid; border-width:0px 0px 0px 10px; border-color:#b8d0d0; margin:20px 0px 10px 10px">
    <h4> Imparare Facendo</h4>
<OL>
<LI>Assegnate alla variabile <b>x</b>  il valore 4/2. Verificate come viene rappresentato <b>x</b>.
<LI>Costruite un numero complesso di parte reale 2.7 e di parte immaginaria -0.73. Calcolatene il complesso coniugato.
<LI>Calcolate il quadrato di 1+1j. Convincetevi della correttezza del risultato.
<LI>Cercate on line informazioni sui numeri ottali e sul modo di rappresentarli in python.
<LI>Costruite la funzione <b>abs_square</b> che prenda come argomento un numero complesso <b>z</b> e restitisca il prodotto di <b>z</b>
      e del suo complesso coniugato. Verificate che il risultato restituito da <b>abs_square</b> coincide, all'interno della precisione dei calcoli fra numeri reali, con <b>abs(z)**2</b>. 
</OL>
</div>

In [14]:
import cmath
x = 4/2
print(x)
z = 2.7 -0.73j
print(z)

2.0
(2.7-0.73j)


##### Informazioni ulteriori

-   Un'introduzione informale ai numeri. [Python tutorial, section 3.1.1](http://docs.python.org/tutorial/introduction.html#using-python-as-a-calculator)

-   Python Library Reference: una presentazione formale dei tipi numerici, <http://docs.python.org/library/stdtypes.html#numeric-types-int-float-long-complex>

-   Think Python, [Sec 2.1](http://www.greenteapress.com/thinkpython/html/book003.html)



3.2 Sequenze
-----------

Stringhe, liste e ntuple (tuples) sono *sequenze*. Possono essere *indicizzate* e *sezionate* (*sliced*) nello stesso modo.

Tuples e stringhe sono “immutable” (fondamentalmente significa che non è possibile cambiare i singoli elementi in una ntupla, e non possiamo modificare i singoli caratteri all'interno di una stringa) mentre le liste sono “mutable” (cioè possiamo cambiare gli elementi di una lista.)

Le sequenze condividono le operazioni seguenti:

<table style="font-size:15px;">
<tr><td>a[i]</td><td>restituisce l'*i*-esimo elemento di `a`; a[0] è il primo elemento</td></tr>
<tr><td>a[i:j]</td><td>restituisce gli elementi da *i* fino a *j* − 1</td></tr>
<tr><td>len(a)</td><td>restituisce il numero di elementi nella sequenza</td></tr>
<tr><td>min(a)</td><td>restituisce il valore più piccolo della sequenza</td></tr>
<tr><td>max(a)</td><td>restituisce il valore più grande nella sequenza</td></tr>
<tr><td>a.count(x)</td><td>restituisce il numero di volte in cui la sottosequenza  x compare nella sequenza a</td></tr>
<tr><td>x in a</td><td>restituisce `True` se `x` è un elemento di `a`, `False` in caso contrario</td></tr>
<tr><td>a + b</td><td>concatena `a` e `b`</td></tr>
<tr><td>n * a</td><td>concatena `n` copie della sequenza `a`</td></tr>
</table>


### 3.2.1 Stringhe

Una stringa è una sequenza (immutable) di caratteri. Una stringa può essere definita usando single quotes:

In [ ]:
a = 'Hello World'

double quotes:

In [ ]:
a = "Hello World"

oppure triple quotes di entrambi i tipi

In [ ]:
a = """Hello World"""
a = '''Hello World'''

Avere più delimitatori possibili per una stringa è utile quando un delimitatore è presente nella stringa. Esempio: supponiamo di voler assegnare alla variabile mystr la stringa *L'ultimo dei Mohicani*. 

In [ ]:
mystr = 'L'ultimo dei Mohicani'
mystr

In [18]:
mystr = "L'ultimo dei Mohicani"
mystr

"L'ultimo dei Mohicani"

La stringa vuota è:

In [ ]:
""

Un particolare carattere può essere estratto usando il suo indice

In [19]:
a = 'trentatre'
a[2]

'e'

In [20]:
my_str = "abc"
print('Il primo carattere di my_str è: ',my_str[0])
print('Il secondo carattere di my_str è: ',my_str[1])

Il primo carattere di my_str è:  a
Il secondo carattere di my_str è:  b


Il numero di caratteri in una stringa (la sua *lunghezza*) si ottiene con la funzione `len()`:

In [21]:
a = "Hello Moon"
print('La lunghezza di "Hello Moon" è:',len(a))
b = "" 
print('La lunghezza della stringa vuota è:',len(b))

La lunghezza di "Hello Moon" è: 10
La lunghezza della stringa vuota è: 0


Si possono combinare (“concatenare”) due stringhe usando l'operatore `+` (è preferibile usare `join` di cui parleremo più avanti):

In [22]:
'Hello ' + 'World'

'Hello World'

Si possono anche concatenare due stringhe semplicemente scrivendole una accanto all'altra:

In [23]:
'a' 'b'

'ab'

Sugli oggetti di Python possono agire metodi (legati a una *classe*), con la sintassi:<br/>
- var.method()

oppure funzioni, che si applicano con la notazione:<br/>
- funzione(var)

Un metodo particolarmente utile è `split()` che converte una stringa in una lista di stringhe:

In [24]:
a = "This is a test sentence."
a.split()

['This', 'is', 'a', 'test', 'sentence.']

Il meodo `split()` divide le stringhe dove trova *white space*. White space significa qualsiasi carattere che viene stampato come spazio bianco, come uno spazio oppure più spazi consecutivi oppure il carattere tab.

È possibile passare una "separator string" a `split()`, spezzando la stringa originaria sulla stringa di separazione (che viene eliminata). Per esempio,
per avere una lista di frasi complete si potrebbe usare il punto "." come separatore:

In [27]:
a = "The dog is hungry. The cat is bored. The snake is awake."
a.split(".")

['The dog is hungry', ' The cat is bored', ' The snake is awake', '']

Oppure:

In [28]:
a = "Pippo(--)Pluto(--)Paperino"
a.split("(--)")

['Pippo', 'Pluto', 'Paperino']

Il metodo opposto per le stringhe è `join` che si usa come segue:

In [30]:
a = "The dog is hungry. The cat is bored. The snake is awake."
s = a.split('.')
s

['The dog is hungry', ' The cat is bored', ' The snake is awake', '']

In [33]:
".".join(s) # join prende una lista come argomento

'The dog is hungry. The cat is bored. The snake is awake.'

In [34]:
" STOP".join(s)  # Si noti che la stringa comincia con uno spazio. 
                 # Lo spazio dopo il separatore è lo spazio che segue il punto nel testo originale.

'The dog is hungry STOP The cat is bored STOP The snake is awake STOP'

Per concatenare stringhe senza introdurre caratteri extra si può utilizzare la `stringa vuota`:

In [36]:
''.join(['Nel',' ','mezzo',' ','del',' ','cammin',' ','di',' ','nostra',' ','vita'])

'Nel mezzo del cammin di nostra vita'

<div style = "width:95%; padding:5px 10px 5px; border-style:solid; border-width:0px 0px 0px 10px; border-color:#b8d0d0; margin:20px 0px 10px 10px">
    <h4> Imparare Facendo</h4>
<OL>
<LI>Trovate il numero di parole nella frase "Quel ramo del lago di Como, che volge a mezzogiorno, tra due catene non interrotte di monti" considerando solo gli spazi come separatori.
<LI>Date le variabili <b>a = 37</b> e <b>b = '42'</b>, fate la somma di <b>a</b> e <b>b</b>
    <OL>
        <LI>come numeri interi.
        <LI>come stringhe.
    </OL>
</OL>
</div>

In [40]:
# 1)

str1 = "Quel ramo del lago di Como, che volge a mezzogiorno, tra due catene non interrotte di monti"
a = str1.split(" ")
len(a)

17

In [44]:
# 2)

a = 37 
b = '42'
c = a + int(b)
z = str(a) + b

print(c)
print(z)

79
3742


Non è possibile modificare direttamente gli elementi di una stringa (è `immutabile`):

In [1]:
a = "This is a test sentence."
a[1] = 3

TypeError: 'str' object does not support item assignment

Ciononostante, le stringhe possiedono parecchi metodi utili che modificano la stringa di partenza. Per esempio `upper()` restituisce una nuova stringa in cui tutti i caratteri della stringa iniziale diventano maiuscoli:

In [52]:
a.upper()

'THIS IS A TEST SENTENCE.'

`a` non viene modificata dall'applicazione di  `upper` che restituisce una nuova stringa:

In [53]:
a

'This is a test sentence.'

È però sempre possibile assegnare la nuova stringa alla variabile precedente:

In [54]:
a = "Going Nowhere Fast"
a = a.upper()
a

'GOING NOWHERE FAST'

<div style = "width:95%; padding:5px 10px 5px; border-style:solid; border-width:0px 0px 0px 10px; border-color:#b8d0d0; margin:20px 0px 10px 10px">
    <h4> Imparare Facendo</h4>
<OL>
<LI>Usate <b>dir</b> per avere la lista di metodi/funzioni applicabili a una stringa.
<LI>Usate <b>help</b> per avere informazioni più dettagliate sulle funzioni <b>index</b>, <b>replace</b> e <b>count</b>.
<LI> Assegnate alla variabile <b>frase</b> il valore <b>"Fratelli d'Italia, l'Italia s'è desta"</b>. In <b>frase</b> determinate l'indice della prima "e"; contate il numero di "i"; sostituite tutti gli apostrofi con "&".
</OL>
</div>

In [2]:
help(str.replace)

Help on method_descriptor:

replace(self, old, new, count=-1, /)
    Return a copy with all occurrences of substring old replaced by new.
    
      count
        Maximum number of occurrences to replace.
        -1 (the default value) means replace all occurrences.
    
    If the optional argument count is given, only the first count occurrences are
    replaced.



In [12]:
a = "Matteo è bellissimo"
print(a.replace("e","i"))

a = "Matteo è bellissimo"
print(a.replace("e","i",1))

Mattio è billissimo
Mattio è bellissimo


In [13]:
help(str.count)

Help on method_descriptor:

count(...)
    S.count(sub[, start[, end]]) -> int
    
    Return the number of non-overlapping occurrences of substring sub in
    string S[start:end].  Optional arguments start and end are
    interpreted as in slice notation.



In [59]:
a = "Matteo è bellissimo"
a.count("t")

2

In [64]:
#dir(str) 
help(str.index)

Help on method_descriptor:

index(...)
    S.index(sub[, start[, end]]) -> int
    
    Return the lowest index in S where substring sub is found,
    such that sub is contained within S[start:end].  Optional
    arguments start and end are interpreted as in slice notation.
    
    Raises ValueError when the substring is not found.



In [83]:
a = "Matteo è bellissimo"
a.index("e",5,)  # string_name.index(cosa_cercare, punto_inizio, punto_fine) 
                 # --> restituisce gli indici delle celle in cui trova tale elemento

10

##### Informazioni ulteriori

-   Introduzione alle stringhe, [Python tutorial 3.1.2](http://docs.python.org/tutorial/introduction.html#strings)



### 3.2.2 Liste

Una liste è una sequenza mutabile di oggetti. Gli  oggetti possono essere di qualsiasi tipo, per esempio numeri interi:

In [ ]:
a = [34, 12, 54]
a

o stringhe:

In [ ]:
a = ['dog', 'cat', 'mouse']
a

Una lista vuota è rappresentata da `[]`:

In [ ]:
a = []

Come per le stringhe, il numero di elementi in una lista può essere trovato con la funzione`len()`:

In [ ]:
a = ['dog', 'cat', 'mouse']
len(a)

È anche possibile *mescolare* tipi diversi nella stessa lista:

In [ ]:
a = [123, 'duck', -42, 17, 0, 'elephant']
a

In Python una lista è un oggetto. È quindi possibile che una lista contenga delle altre liste perchè una lista contiene una sequenza di oggetti:

In [ ]:
a = [1, 4, 56, [5, 3, 1], 300, 400]

Si possono combinare (“concatenare”) due liste usando l'operatore `+`:

In [ ]:
[3, 4, 5] + [34, 35, 100]

Si possono modificare gli elementi di una lista usando una assegnazione:

In [ ]:
a = [1,2]
a

In [ ]:
a = ["pippo",3.14,"pluto"]
a

Per accedere agli elementi di una lista:

In [ ]:
a = ['a','b','c','d','e','f']
print(a[1])
print(a[2:4])

In [ ]:
a[2] = 'f'
print(a)

Si può aggiungere un oggetto al fondo di una lista con il metodo `append()`. Notate che `append` modifica la lista su cui agisce e non restituisce nulla; come si dice, agisce "in place".

In [ ]:
a = [34, 56, 23]
a.append(42)
a

Si può inserire un oggetto all'indice voluto in una lista con il metodo `insert()`:

In [ ]:
aList = [123, 'xyz', 'zara', 'abc']
aList.insert( 3, 2009)
print("Final List : ", aList)

Si può eliminare un oggetto da una lista chiamando il metodo `remove()` e passandogli l'oggetto da eliminare. Per esempio:

In [ ]:
a = [34, 56, 23, 42]
a.remove(56)
a

Ad una lista posso applicare il metodo `pop`, che non ho menzionato. Per ottenere informazioni più dettagliate:

In [14]:
help(list.pop)  

Help on method_descriptor:

pop(self, index=-1, /)
    Remove and return item at index (default last).
    
    Raises IndexError if list is empty or index is out of range.



#### Nested lists
Gli elementi di una lista possono, a loro volta, essere liste in questo caso si parla di <b>liste annidate</b> o <b>nested lists</b>.

In [21]:
a_nest = [10,[20,21,22],30]

Per accedere agli elementi di una nested list:

In [30]:
a_nest[1]  #stampa l'elemento di indice 1 (in questo caso una lista a sua volta)

[20, 21, 22]

In [31]:
a_nest[1][2]  #elemento di indice 2 dell'elemento 1 (ovvero di una lista)

22

#### Come ordinare una lista

Posso usare il metodo `sort` oppure la funzione `sorted`:

In [1]:
l = [1,3,2]
dir(l)

['__add__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__rmul__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'append',
 'clear',
 'copy',
 'count',
 'extend',
 'index',
 'insert',
 'pop',
 'remove',
 'reverse',
 'sort']

Il metodo `sort` modifica la lista su cui opera e NON restituisce una lista risultato.

In [2]:
l.sort()
print(l)

[1, 2, 3]


La funzione `sorted` non modifica la lista su cui viene fatta agire e restituisce una nuova lista

In [3]:
l1=sorted(l)
print(l1)

[1, 2, 3]


#### Il comando range()

Il comando `range(n)` genera la lista degli interi da 0 fino a *n-1* (attenzione!), un tipo speciale di lista, che è spesso necessaria. Alcuni esempi:

In [39]:
list(range(3))

[0, 1, 2]

In [40]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

Questo comando si usa spesso con i `for loops`. Per esempio, per stampare i numeri 0<sup>2</sup>,1<sup>2</sup>,2<sup>2</sup>,3<sup>2</sup>,…,10<sup>2</sup>, si può usare il codice seguente:

In [4]:
for i in range(11):
    print(i**2)

0
1
4
9
16
25
36
49
64
81
100


Il comando range prende un parametro opzionale per l'inizio della sequenza di interi  (start) e un altro parametro opzionale per il *passo* (*step*). Spesso si scrive `range([start],stop,[step])` dove gli argomenti in parentesi quadrate (square brackets) (cioè start e step) sono opzionali. 
Notate che il valore corrispondente a `start` è il primo elemento presente nella lista, mentre il valore corrispondente a `stop` è il "primo escluso" e non compare nella lista. Alcuni esempi:

In [ ]:
list(range(3, 10))            # start=3

In [ ]:
list(range(3, 10, 2))         # start=3, step=2

In [ ]:
list(range(10, 0, -1))        # start=10,step=-1

In [5]:
list(range(0.,1.,0.1))

TypeError: 'float' object cannot be interpreted as an integer

Perchè usiamo `list(range())`?

In Python 3, `range()` genera i numeri on demand. Quando si usa `range()` in un for loop, questo è più efficiente, perchè non occupa memoria con una lista di numeri. Passare il comando `range()` a `list()` lo forza a generare immediatamente tutti i numeri.

<div style = "width:95%; padding:5px 10px 5px; border-style:solid; border-width:0px 0px 0px 10px; border-color:#b8d0d0; margin:20px 0px 10px 10px">
    <h4> Imparare Facendo</h4>
    <OL>
<LI>Assegnate alla variabile <b>L1</b> la lista [1,4,2,3].
    <OL> 
        <LI>Create una nuova lista <b>L2</b> ordinando, in ordine crescente, <b>L1</b> con il metodo <b>sorted</b>.
        <LI>Ordinate in ordine decrescente la lista <b>L1</b> con il metodo <b>sorted</b>.
        <LI>Eliminate l'ultimo elemento di <b>L1</b>.
    </OL>
<LI>Costruite la lista degli interi decrescenti da 9 a 2 entrambi compresi.
<LI>Costruite la lista dei numeri divisibile per 3 compresi fra 0 e 100.
</OL>
</div>

<div style = "width:95%; padding:5px 10px 5px; border-style:solid; border-width:0px 0px 0px 10px; border-color:#b8d0d0; margin:20px 0px 10px 10px">
    <h4> Imparare Facendo</h4>
<OL>
<LI>Usate <b>dir</b> per avere la lista di metodi/funzioni applicabili a una lista.
<LI>Usate <b>help</b> per avere informazioni più dettagliate sulle funzioni <b>extend</b>, e <b>reverse</b>.
    Notate che entrambi i metodi agiscono "in place"
<LI>Assegnate alla variabile <b>lis</b> il valore [1,2,3].
    <OL> 
        <LI>Moltiplicate <b>lis</b> per 2.
        <LI>Costruite a partire da <b>lis</b> la lista <b>lis1</b> con gli elementi di <b>lis</b> in ordine inverso.
        <LI>Costruite la somma di <b>lis</b> e <b>lis1</b> usando il comando <b>append</b>.
        <LI>Costruite la somma di <b>lis</b> e <b>lis1</b> usando il comando <b>extend</b>.
    </OL>
<LI>Costruite una funzione <b>add_to_list</b> che prenda come argomento una lista <b>L</b> e la restituisca aggiungendo come ultimi elementi <b>"penultimo"</b> e <b>"ultimo"</b>.
<LI>Costruite una funzione <b>reverse</b> che prenda come argomento una lista di numeri reali <b>L</b> e la restituisca ordinata.
</OL>
</div>

##### Informationi ulteriori

-   Introduzione alle Liste, [Python tutorial, section 3.1.3](http://docs.python.org/tutorial/introduction.html#lists)
- Funzioni e metodi che operano su Liste, [Python tutorial, section 5.1](https://docs.python.org/3/tutorial/datastructures.html)

### 3.2.3 Tuples

A *ntupla* (*tuple*) è una sequenza (immutable) di oggetti. Le ntuple si comportano in modo molto simile alle liste con l'eccezione che non possono essere modificate.

Per esempio, qualunque tipo di oggetto vi può comparire:

In [129]:
a = (12, 13, 'dog')
a

(12, 13, 'dog')

In [130]:
a[0]

12

Le parentesi non sono necessarie per definire una tuple: una sequenza di oggetti separati da virgole è sufficiente a definire una tuple:

In [131]:
a = 100, 200, 'duck'
a

(100, 200, 'duck')

benchè sia buona pratica includere le parantesi per una maggiore leggibilità.

Si può costruire una ntuple con applicando il comando <b>tuple</b> ad una sequenza:

In [128]:
tuple([1,'a'])

(1, 'a')

In [133]:
tuple("ciao")

('c', 'i', 'a', 'o')

Tuples possono anche essere usate per assegnare due valori contemporameamente:

In [134]:
x, y = 10, 20
x

10

In [43]:
y

20

Le ntuple possono essere usate per scambiare due oggetti in una sola riga di codice. Per esempio:

In [135]:
x = 1
y = 2
x, y = y, x
x

2

In [45]:
y

1

La ntupla vuota è `()`.

In [138]:
t = ()
len(t)

0

In [139]:
type(t)

tuple

La notazione per una ntupla contenente un solo valore può, all'inizio, sembrare un po' strana:

In [48]:
t = (42,)
type(t)

tuple

In [49]:
len(t)

1

La virgola addizionale è necessaria per distinguere `(42,)` da `(42)`. Nel secondo caso le parentesi indicherebbero solo l'ordine di precedenza delle operazioni: `(42)` si semplifica a `42` che è semplicemente un numero:

In [50]:
t = (42)
type(t)

int

Un esempio dell'immutabilità di una tuple:

In [143]:
a = (12, 13, 'dog')
a[2]

'dog'

In [54]:
a[0] = 1

TypeError: 'tuple' object does not support item assignment

Operazioni possibili sulle ntuple

In [140]:
dir(a)

['__add__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'count',
 'index']

In [141]:
len(a)

3

In [144]:
a.index(13)

1

L'immutabilità è la differenza principale fra una tuple e una lista (che è mutable). Le tuples vengono utilizzate quando non si vuole che il contenuto venga modificato.

Le funzioni Python che ritornano più di un valore, li ritornano in tuples (che è sensato perchè in genere non si vuole che i risultati di una chiamata a una funzione vengano modificati).

<div style = "width:95%; padding:5px 10px 5px; border-style:solid; border-width:0px 0px 0px 10px; border-color:#b8d0d0; margin:20px 0px 10px 10px">
    <h4> Imparare Facendo</h4>
<OL>
    <LI> Costruite la <b>lista</b> dei numeri divisibili per 5 nell'intervallo 130-180, estremi inclusi. Assegnate la lista alla variabile <b>L0</b>. 
    <OL>
        <LI> Trasformate <b>L0</b> in una ntupla.
        <LI> Determinate l'indice di 145 in <b>L0</b>
        <LI> Sommate <b>L0</b> a sè stessa
</OL>
</div>

In [161]:
L0 = list(range(130,185,5))
print(L0)
L0 = tuple(L0)
print('\nIndice elemento 145:', L0.index(145))
L0 = L0+L0
print('\n',L0)

[130, 135, 140, 145, 150, 155, 160, 165, 170, 175, 180]

Indice elemento 145: 3

 (130, 135, 140, 145, 150, 155, 160, 165, 170, 175, 180, 130, 135, 140, 145, 150, 155, 160, 165, 170, 175, 180)


### Indici in sequenze

Abbiamo già accennato più volte all'uso degli indici in sequenze introducendo le varie strutture dati. Questa sezione cerca di sistematizzare l'argomento.<BR>&nbsp;<BR>
I singoli oggetti in una lista/stringa/tuple possono essere indicati con il loro indice fra parentesi quadre(`[` e `]`). Notate che Python (come in C ma diversamente dal Fortran e da Matlab) conta gli indici partendo da zero!

In [162]:
a = ['dog', 'cat', 'mouse']
a[0]

'dog'

In [163]:
a[1]

'cat'

In [164]:
a[2]= 'parrot'

In [165]:
a

['dog', 'cat', 'parrot']

L'indice di un oggetto in una sequenza si può trovare con il metodo `index`(oppure `find`, cercate sulladocumentazione la differenza fra i due metodi):

In [ ]:
a.index('cat')

In [168]:
b = [0,'pluto',10,3]
print(b.index(10))
type(b)

2


list

<div style= 'padding:20px 20px 0px 150px;'>
    <img src="../Humour/FirstWord.jpg" width="500" align="center" margin="0px 0px 0px 50px"/>
</div>

Python ha un modo semplice per estrarre l'ultimo elemento di una lista: si usa l'indice “-1”, dove il segno meno indica che l'elemento è a *un passo* dal fondo della lista. Nello stesso, l'indice “-2” ritorna il penultimo elemento (2nd last):

In [169]:
a = ('dog', 'cat', 'mouse')
a[-1]

'mouse'

In [17]:
a[-2]

'cat'

Si può anche pensare ad `a[-1]` come una abbreviazione di `a[len(a) - 1]`.

Ricordate che le stringhe (come le liste) sono delle sequenze e possono essere indicizzate nello stesso modo:

In [170]:
a = "Hello World!" 
a[0]

'H'

In [171]:
a[1]

'e'

In [172]:
a[100]

IndexError: string index out of range

In [173]:
a[-1]

'!'

In [174]:
a[-2]

'd'

In [176]:
a[3]='u'

TypeError: 'str' object does not support item assignment

##### Informazioni ulteriori

-   Introduzione a stringhe e indici [Python tutorial, section 3.1.2](http://docs.python.org/tutorial/introduction.html#strings), la sezione sull'uso degli indici inizia dopo che sono state introdotte le stringhe.

### Sezioni di sequenze

Lo *slicing* di una sequenze può essere utilizzato per estrarre più di un elemento. Per esempio:

In [177]:
a = "Hello World!"
a[5:10]

' Worl'

Scrivendo `a[0:3]` richiediamo i primi 3 elementi iniziando dall'elemento 0. Analogamente:

In [178]:
a[1:4]

'ell'

In [179]:
a[0:2]

'He'

In [180]:
a[0:6]

'Hello '

Si possono usare indici negtivi per riferirsi alla fine della sequenza:

In [181]:
a[0:-1]

'Hello World'

Per arrivare fino alla fine della sequenza:

In [182]:
a[5:len(a)]

' World!'

È anche possibile omettere l'indice inziale o quello finale e questo restitusce tutti gli elementi a partire dall'inizio oppure fino alla fine della sequenza. Qualche esempio per chiarire:

In [183]:
a = "Hello World!"
a[:5]

'Hello'

In [33]:
a[5:]

' World!'

In [184]:
a[-2:]

'd!'

In [185]:
a[:]

'Hello World!'

Per invertire l'ordine degli elementi di una lista possiamo usare:

In [186]:
a[::-1]

'!dlroW olleH'

Notate che `a[:]` genera una *copia* di `a`. Il modo con cui vengono usati gli indici nello slicing può sembrare controintuitivo. Se vi trovate in difficoltà, ricordate questa spiegazione dal [Python tutorial (section 3.1.2)](http://docs.python.org/tutorial/introduction.html#strings):

> The best way to remember how slices work is to think of the indices as pointing between characters, with the left edge of the first character numbered 0. Then the right edge of the last character of a string of 5 characters has index 5, for example:
>
>        0   1   2   3   4     <-- use for INDEXING 
>       -5  -4  -3  -2  -1     <-- use for INDEXING from the end
>      +---+---+---+---+---+           
>      | H | e | l | l | o |
>      +---+---+---+---+---+ 
>      0   1   2   3   4   5   <-- use for SLICING
>     -6  -5  -4  -3  -2  -1   <-- use for SLICING from the end
>                                      
>
> The first row of numbers gives the position of the slicing indices 0...5 in the string; the second row gives the corresponding negative indices. The slice from i to j consists of all characters between the edges labelled i and j, respectively.

L'affermazione importante è che nello *slicing* gli indici puntano *fra* i caratteri o elementi.

Per *indicizzare* è preferibile pensare che gli indici si riferiscano direttamente agli elementi. 

Se non vi ricordate esattamente come funzionano gli indici, potete semplicemente sperimentare con un piccolo esempio usando il Python prompt prima o durante la scrittura del vostro programma.

In [188]:
a = "Hello"
a[-3:-6:-1]

''

La posizione di un elemento in qualunque oggetto di tipo sequenza può essere trovato con il *metodo* index:

In [41]:
(1,2,3).index(2)

1

Se l'elemento cercato non è presente si ha un errore:

In [42]:
[1,2,3].index(4)

ValueError: 4 is not in list

<div style = "width:95%; padding:5px 10px 5px; border-style:solid; border-width:0px 0px 0px 10px; border-color:#b8d0d0; margin:20px 0px 10px 10px">
    <h4> Imparare Facendo</h4>
<OL>
<LI>Assegnate alla variabile <b>L0</b> la lista di tutti i numeri fra 175 e 227, 175 incluso e 227 escluso.
    <OL>
        <LI>Estraete da <b>L0</b> il penultimo elemento.
        <LI>Estraete da <b>L0</b> gli elementi di indice 13, 14 e 15.
        <LI>Estraete da <b>L0</b> tutti gli elementi salvo il primo e l'ultimo.
    </OL>
<LI> Data la stringa <b>"Supercalifragilisticexpiralidoso"</b> costruite la stringa con tutte le lettere in ordine inverso.
<LI>Costruite una funzione <b>reverse</b> che prenda come argomento una lista di numeri reali <b>L</b> e la restituisca ordinata inversamente, dal più grande al più piccolo.
</OL>
</div>

##### Informazioni ulteriori

-   Introduzione a stringhe, indici e sezioni [Python tutorial, section 3.1.2](http://docs.python.org/tutorial/introduction.html#strings)



## 3.3 Dizionari

I dizionari sono anche chiamati “associative arrays” o “hash tables”. I dizionari sono insiemi *unordered* di coppie *key-value*. Le keywords devono essere oggetti immutabili mentre i dizionari sono oggetti mutabili.

Un dizionario vuoto può essere creato usando parentesi graffe (curly braces):

In [19]:
d = {}
print(d)

{}


Coppie keyword-value possono essere aggiunti come segue:

In [20]:
d['today'] = '22 deg C'    # 'today' è la keyword

In [21]:
d['yesterday'] = '19 deg C'

In [22]:
print(d)

{'today': '22 deg C', 'yesterday': '19 deg C'}


Il valore può essere estratto usando la keyword come indice:

In [90]:
d['today']

'22 deg C'

Altri modi di riempire un dizionario se i dati sono noti al momento in cui viene creato sono:

In [23]:
d2 = {2:4, 3.5:9, 4:16, 5:25}   # Notate che sto usando dei numeri interi o reali come chiavi.
d2

{2: 4, 3.5: 9, 4: 16, 5: 25}

In [24]:
d2_a = {'p':4, 'q':9, 'r':16, 's':25}   # Se uso stringhe come chiavi devo scriverli fra apici
d2_a

{'p': 4, 'q': 9, 'r': 16, 's': 25}

oppure:

In [26]:
d3 = dict(a=1, b=2, c=3)     # In questo caso, se uso stringhe come chiavi NON devo scriverli fra apici
                             # In questo caso, non posso usare numeri interi o reali come chiavi.
d3

{'a': 1, 'b': 2, 'c': 3}

Morale della favola: se esistono più modi di dare un comando non è garantito che ci sia una ovvia coerenza fra i diversi modi.

La funzione `dict()` crea un dizionario vuoto.<BR>
Per eliminare una coppia key-value da un dizionario si può usare `del` passando la key come argomento:

In [91]:
dd = {'p':4, 'q':9, 'r':16, 's':25}
print(dd)

{'p': 4, 'q': 9, 'r': 16, 's': 25}


In [92]:
del dd['r']
print(dd)

{'p': 4, 'q': 9, 's': 25}


`d.keys()` restituisce la lista di tutte le keys:

In [93]:
d.keys()

dict_keys(['today', 'yesterday'])

Altri metodi utili per i dizionari sono `values()`, `items()` e `get()`. Si può usare `in` per controllare se un certo valore è presente.

In [94]:
type(d.values())

dict_values

In [95]:
d.items()

dict_items([('today', '22 deg C'), ('yesterday', '19 deg C')])

Il metodo `get(key,default)` restituisce il valore corrispondente a una certa `key` se la key esiste, altrimenti restituisce l'oggetto di `default`.

In [96]:
d.get('today','unknown')

'22 deg C'

In [98]:
d.get('tomorrow','culo')

'culo'

In [99]:
'today' in d

True

In [100]:
'tomorrow' in d

False

Ulteriori dettagli:

-   La keyword può essere qualsiasi (immutable) oggetto in Python. Questo include:

    -   numeri

    -   stringhe

    -   tuples.

-   i dizionari restituiscono i valori molto velocemente quando viene loro fornita la key.

In [83]:
vlist = [1.2,-0.3,3.5]

Componente `y`:

In [84]:
vlist[1]

-0.3

Rappresentazione con un dizionario:

In [85]:
vdict = {'vx':1.2, 'vy':-0.3, 'vz':3.5}

Componente `y`:

In [86]:
vdict['vy']

-0.3

I valori di un dizionario possono essere a loro volte dizionari (oppure liste e ntuple):

In [27]:
nazioni = {'Francia':{'capitale':'Parigi','popolazione':67.39e6},
           'Germania':{'capitale':'Berlino','popolazione':83.24e6}}

In [28]:
nazioni['Germania']

{'capitale': 'Berlino', 'popolazione': 83240000.0}

In [102]:
nazioni['Germania']['popolazione']

83240000.0

<div style = "width:95%; padding:5px 10px 5px; border-style:solid; border-width:0px 0px 0px 10px; border-color:#b8d0d0; margin:20px 0px 10px 10px">
    <h4> Imparare Facendo</h4>
<OL>
    <LI>Costruite il dizionario <b>squares</b> che abbia come keys gli interi fra 1 e 5 e come values i corrispondenti quadrati.
        <OL>
            <LI>Aggiungete a <b>squares</b> la coppia key:value {6:36}.
            <LI>Eliminate da <b>squares</b> la coppia key:value {1:1}.
        </OL>
    <LI>Costruite il dizionario <b>mesi</b> che abbia come keys il nome dei mesi e come values il corrispondente numero di giorni
        sotto forma di stringa. Per febbraio il value sia "28 oppure 29".
</OL>
</div>

In [109]:
squares = {1:1, 2:4, 3:9, 4:16, 5:25}
squares[6]=36
del squares[1]
print(squares)

{2: 4, 3: 9, 4: 16, 5: 25, 6: 36}


In [114]:
mesi = dict(gennaio='31', febbraio='28,29', marzo='31', aprile='30', maggio='31', giugno='30', luglio='31', agosto='31', settembre='30', ottobre='31', novembre='31', dicembre='31')
print(mesi)

{'gennaio': '31', 'febbraio': '28,29', 'marzo': '31', 'aprile': '30', 'maggio': '31', 'giugno': '30', 'luglio': '31', 'agosto': '31', 'settembre': '30', 'ottobre': '31', 'novembre': '31', 'dicembre': '31'}


## 3.4 Conversioni fra tipi

La funzione `type` mostra il tipo di un oggetto. È possibile convertire un oggetto in uno di tipo diverso.

La funzione `int()` converte in qualunque oggetto trasformabile in un intero:

In [ ]:
a = '34'       # a è una stringa che contiene i caratteri 3 e 4
x = int(a)     # x è il numero intero 34

In [116]:
print("Il type di '34' è",type(a))
print("Il type di",x," è",type(x))

Il type di '34' è <class 'list'>
Il type di 2.0  è <class 'float'>


Ovviamente l'argomento di `int` deve essere interpretabile come un intero

In [115]:
int('piripacchio')

ValueError: invalid literal for int() with base 10: 'piripacchio'

Agendo su numeri reali `int()` tronca qualsiasi parte non intera:

In [45]:
int(7.0)

7

In [44]:
int(7.9)

7

In [43]:
int(-7.9)

-7

Per arrotondare un numero reale all'intero più vicino usate la funzione `round()`:

In [117]:
round(7.9)

8

In [118]:
round(-7.9)

-8

Il tipo di una stringa è `str`. La funzione `str` trasforma un oggetto in stringa:

In [119]:
a = "Hello World"
type(a)

str

In [120]:
b = ""
type(b)

str

In [53]:
num = 5.76
print(type(num))

<class 'float'>


In [54]:
num_as_string = str(num)
print("num_as_string is of type",type(num_as_string))

num_as_string is of type <class 'str'>


Il tipo di una lista è `list`:

In [121]:
a = [34, 12, 54]
type(a)

list

In [122]:
type([])

list

## 3.5 Locazione in memoria e copia di oggetti

Python fornisce la  funzione `id()`  che restituisce un numero che identifica in modo unico un oggetto, essenzialmente l'indirizzo in memoria. Possiamo utilizzare questa funzione per verificare se due oggeti sono identici, cioè se si riferiscono alla stessa locazione di memoria.<BR>

In [123]:
m = 2
print(f'id(m) = {id(m)}')

id(m) = 140718087967560


Quando eseguiamo il comando `n = m`, la variabile `n` punta alla stessa posizione di memoria:

In [124]:
n = m
print(f'id(n) = {id(n)}')

id(n) = 140718087967560


Usando `id`

In Python, i numeri, le ntuple, le stringhe sono "immutabili" cioè non possono essere modificati senza creare una nuova copia dell'oggetto.<br>
Se cambiamo il valore di n cambia anche la posizione in memoria:

In [125]:
n += 1
print(f'n = {n}')
print(f'id(n) = {id(n)}')

n = 3
id(n) = 140718087967592


L'indirizzo che è immagazzinato nella variabile m, invece, non cambia:

In [64]:
print(f'm = {m}')
print(f'id(m) = {id(m)}')

m = 2
id(m) = 140718087967560


Copia di liste tramite assegnazione. Qualche esempio:

In [65]:
a = list(range(10))
a

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [66]:
b = a           # shallow copy. b punta allo stesso oggetto a cui punta a
b[0] = 42
a               # cambiare b cambia anche a

[42, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [67]:
id(a)

1751581580288

In [68]:
id(b)

1751581580288

In [69]:
id(b[0])

140718087968840

Per creare una nuova copia indipendente di una sequenza di oggetti (incluse le liste), possiamo usare lo slicing. Se `a` è una lista, allora `a[:]` restituisce una copia indipendente di `a`.

In [70]:
c = a[:] 
id(c)          # c è un oggetto diverso

1751599071424

In [126]:
c[0] = 100       
a              # cambiare c non cambia a

[34, 12, 54]

In [127]:
c              # verifichiamo c

[100, 1, 2, 3, 4, 5, 6, 7, 8, 9]

La standard library di Python fornisce il modulo `copy`, che contiene funzioni che possono essere utilizzate per creare copie di oggetti. Avremmo potuto usare `import copy; c = copy.deepcopy(a)` invece di `c = a[:]`.

### 3.5.1 Mutabilità, Immutabilità e metodi che modificano oggetti immutabili

In [73]:
st = "abc"

Non posso sostitire una lettera in st

In [74]:
st[0] = 'A'

TypeError: 'str' object does not support item assignment

Posso però fare

In [ ]:
st = st.upper()
st

Il metodo upper crea una nuova stringa. Quando la assegno a `st` distruggo la referenza all'oggetto originale.

In [75]:
st = "abc"
print(id(st))

140718086767680


In [76]:
st = st.upper()
print(id(st))

1751586164720


La stessa cosa succede con ntuple e numpy.array.

Se modifico un oggetto mutabile la sua locazione in memoria non viene cambiata.

In [79]:
lst = [1,2,3]
print(lst,id(lst))
lst[0] = 100
print(lst,id(lst))
lst.append(44)
print(lst,id(lst))

[1, 2, 3] 1751582155648
[100, 2, 3] 1751582155648
[100, 2, 3, 44] 1751582155648


<div style= 'padding:20px 20px 0px 200px;'>
    <img src="../Humour/Novice_vs_Experienced_Programmer.jpg" width="300" align="center"/>
</div>